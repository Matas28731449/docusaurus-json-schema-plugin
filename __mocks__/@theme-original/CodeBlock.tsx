import React, { ReactNode } from "react"

import type { JSX } from "react"

type Props = {
  children: ReactNode
  language?: string
}

export default function CodeBlock({ children }: Props): JSX.Element {
  // Keep it simple, as lib is not using advanced feature
  let rawString =
    typeof children === "string" ? children : JSON.stringify(children)

  return <code>{rawString}</code>
}
