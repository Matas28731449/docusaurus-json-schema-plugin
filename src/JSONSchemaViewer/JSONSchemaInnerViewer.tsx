import React from "react"

import type { JSONSchema7 } from "json-schema"

type InnerViewerProperties = {
  schema: JSONSchema7
}

function JSONSchemaInnerViewer(props: InnerViewerProperties): JSX.Element {
  return <></>
}

export default JSONSchemaInnerViewer
