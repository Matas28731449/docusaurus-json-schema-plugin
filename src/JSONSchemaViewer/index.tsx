import React, { useEffect, useState } from "react"
import Layout from "@theme/Layout"
import Translate from "@docusaurus/Translate"
import AllOfMerger from "json-schema-merge-allof"
import { Resolver } from "@stoplight/json-ref-resolver"

import type { JSONSchema7 } from "json-schema"

import JSONSchemaViewer from "./JSONSchemaInnerViewer"

type ViewerProperties = {
  loadSchema: () => Promise<JSONSchema7>
}

// merged representation of allOf array of schemas
async function mergeAllOf(
  schema: JSONSchema7
): Promise<Omit<JSONSchema7, "allOf">> {
  const mergedSchema = AllOfMerger(schema, {
    // Technically valid, but only interested on fields with name, at least for now ...
    ignoreAdditionalProperties: true,
  })

  return Promise.resolve(mergedSchema as Omit<JSONSchema7, "allOf">)
}

function JSONSchemaEditor(props: ViewerProperties): JSX.Element {
  const [schema, setSchema] = useState(
    undefined as undefined | Omit<JSONSchema7, "allOf">
  )
  const [fetchError, setFetchError] = useState(undefined as undefined | Error)

  useEffect(() => {
    props
      .loadSchema()
      .then((userSchema) => new Resolver().resolve(userSchema))
      .then((resolvedSchema) => mergeAllOf(resolvedSchema.result))
      .then((schemaWithoutAllOf) => setSchema(schemaWithoutAllOf))
      .catch((err) => setFetchError(err))
  }, [])

  return (
    <Layout>
      {fetchError !== undefined && (
        <div>
          <p>
            <Translate
              values={{
                errorMessage: fetchError.message,
                id: "docusaurus-json-schema-viewer-plugin.error",
              }}
            >
              {"This component crashed because of error: {errorMessage}."}
            </Translate>
          </p>
        </div>
      )}
      {schema === undefined && (
        <div>
          <p>
            <Translate values={{ id: "docusaurus-json-schema-viewer-plugin.loading" }}>
              {"Loading schema ..."}
            </Translate>
          </p>
        </div>
      )}
      {schema !== undefined && <JSONSchemaViewer schema={schema} />}
    </Layout>
  )
}

export default JSONSchemaEditor
