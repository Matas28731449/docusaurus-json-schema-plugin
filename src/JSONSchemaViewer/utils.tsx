import React from "react";

import type { JSONSchema7 } from "json-schema";

// To keep the same type everywhere here
type cleanSchema = Omit<JSONSchema7, "allOf">;

