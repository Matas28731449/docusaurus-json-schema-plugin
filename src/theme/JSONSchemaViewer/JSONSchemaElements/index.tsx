// Component to deal with oneOf / anyOf / allOf / not
export { SchemaComposition } from "./schemaComposition/index"

// Component to deal with if-then-else , dependentRequired , dependentSchemas , dependencies
export { SchemaConditional } from "./SchemaConditional/index"

// Component to deal with Object
export { CreateObject } from "./object/index"

// Component to deal with Array
export { CreateArray } from "./array/index"

// Component to deal with String
export { default as CreateString } from "./CreateString"

// Component to deal with Boolean
export { default as CreateBoolean } from "./CreateBoolean"

// Component to deal with Number
export { default as CreateNumber } from "./CreateNumber"

// Component to deal with Integer
export { default as CreateInteger } from "./CreateInteger"

// Component to deal with schema defined as "true"
export { default as CreateAlwaysValid } from "./CreateAlwaysValid"

// Component to deal with schema defined as "false"
export { default as CreateAlwaysInvalid } from "./CreateAlwaysInvalid"
