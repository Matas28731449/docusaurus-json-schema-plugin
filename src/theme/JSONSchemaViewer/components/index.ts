// To avoid circular dependancies issue, let's have a single place for exports and imports

// Collapsible componenent
export { default as Collapsible } from "./Collapsible"

// CreateDetailsNode component
export { default as CreateDetailsNode } from "./CreateDetailsNode"

// CreateEdges component
export { default as CreateEdges } from "./CreateEdges"

// CreateProperties component
export { default as CreateProperties } from "./CreateProperties"

// RenderAnyOneOf component
export { default as RenderAnyOneOf } from "./RenderAnyOneOf"

// https://stackoverflow.com/a/69328045/6149867
export type WithRequired<T, K extends keyof T> = T & { [P in K]-?: T[P] }
