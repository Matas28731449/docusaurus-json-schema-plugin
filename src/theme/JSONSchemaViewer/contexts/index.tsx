export * from "@theme/JSONSchemaViewer/contexts/jsvOptions"
export * from "@theme/JSONSchemaViewer/contexts/schemaHierarchy"
