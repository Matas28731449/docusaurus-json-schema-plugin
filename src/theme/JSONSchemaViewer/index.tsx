import React, { useState, useEffect } from "react"
import { Resolver } from "@stoplight/json-ref-resolver"

import { CreateNodes, Collapsible } from "./components/index"
import { JSVOptionsContextProvider } from "./contexts/index"

import type { JSONSchema } from "./types"
import type { JSVOptions } from "./contexts/index"
import type { IResolveOpts } from "@stoplight/json-ref-resolver/types"

export type Props = {
  // The schema to use
  schema: unknown
  // To customize the ref resolving
  resolverOptions?: IResolveOpts
  // To customize the viewer
  viewerOptions?: Omit<JSVOptions, "fullSchema">
  // Other properties (who knows ?)
  [x: string]: any
}

type InnerViewerProperties = {
  // Thanks to @stoplight/json-ref-resolver, $ref are either :
  // 1. resolved
  // 2. unresolved (as circular stuff are not on the roadmap)
  schema: JSONSchema
  // Options for viewer
  viewerOptions?: Omit<JSVOptions, "fullSchema">
}

// Internal
function JSONSchemaInnerViewer(props: InnerViewerProperties): JSX.Element {
  const { schema, viewerOptions } = props
  // Title of the schema, for user friendliness
  const title =
    typeof schema !== "boolean" && schema?.title !== undefined
      ? schema.title
      : "Schema"

  // state for provider
  const startingState: JSVOptions = {
    fullSchema: schema,
    // spread provided attributes
    ...viewerOptions,
  }

  return (
    <JSVOptionsContextProvider value={startingState}>
      <Collapsible
        summary={<strong>{title}</strong>}
        detailsProps={{
          open: true,
        }}
      >
        <CreateNodes schema={schema} />
      </Collapsible>
    </JSVOptionsContextProvider>
  )
}

// Entry point
export default function JSONSchemaViewer(props: Props): JSX.Element {
  const { schema: originalSchema, resolverOptions, viewerOptions } = props

  const [error, setError] = useState(undefined as undefined | Error)
  const [resolvedSchema, setResolvedSchema] = useState(
    undefined as undefined | JSONSchema
  )

  useEffect(() => {
    // Time to do the job
    new Resolver()
      .resolve(originalSchema, resolverOptions)
      .then((result) => {
        setResolvedSchema(result.result)
      })
      .catch((err) => {
        setError(err)
      })
  }, [])

  if (error !== undefined) {
    return <div>Something bad happens : {error.message}</div>
  } else if (resolvedSchema === undefined) {
    return <div>Loading ....</div>
  } else {
    return (
      <JSONSchemaInnerViewer
        schema={resolvedSchema}
        viewerOptions={viewerOptions}
      />
    )
  }
}
