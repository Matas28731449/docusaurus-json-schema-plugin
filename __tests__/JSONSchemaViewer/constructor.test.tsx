import React from "react"

// For typings autocomplete whatever your IDE
import { expect, test, describe } from "@jest/globals"

import { create, act } from "react-test-renderer"

import JSONSchemaViewer from "../../src/theme/JSONSchemaViewer/index"

// Type to prevent creating invalid mocks
import type { JSONSchema } from "../../src/theme/JSONSchemaViewer/types"

// Type for react-test-renderer
import type { ReactTestRenderer } from "react-test-renderer"
import CodeBlock from "../../__mocks__/@theme-original/CodeBlock"

describe("JSONSchemaViewer - constructor", () => {
  test("Overwrite default qualifierMessagesOrder value", async () => {
    const fakeSchema: JSONSchema = {
      type: "object",
      minProperties: 1,
    }

    // render the component
    let root: ReactTestRenderer | undefined
    await act(async () => {
      root = create(
        <JSONSchemaViewer
          schema={fakeSchema}
          viewerOptions={{ qualifierMessagesOrder: ["objectProperties"] }}
        />,
      )
    })

    // make assertions on root
    expect(root?.toJSON()).toMatchSnapshot()
  })

  test("Overwrite default DescriptionComponent value", async () => {
    const fakeSchema: JSONSchema = {
      type: "object",
      // Markdown text
      description: "# Hello, *world*!",
    }

    // render the component
    let root: ReactTestRenderer | undefined
    await act(async () => {
      root = create(
        <JSONSchemaViewer
          schema={fakeSchema}
          viewerOptions={{
            // To simulate "react-markdown" like libraries
            // In prod context, it will likely be invoked like this
            // DescriptionComponent: ({description}) => <ReactMarkdown children={description} />
            DescriptionComponent: () => (
              <h1>
                {" "}
                Hello, <em>world</em>!{" "}
              </h1>
            ),
          }}
        />,
      )
    })

    // make assertions on root
    expect(root?.toJSON()).toMatchSnapshot()
  })

  test("Overwrite default UnresolvedRefsComponent value", async () => {
    // tree schema, extensible
    // https://json-schema.org/draft/2020-12/release-notes#dynamicref-and-dynamicanchor
    const fakeSchema: JSONSchema = {
      $schema: "https://json-schema.org/draft/2020-12/schema",
      $id: "https://example.com/tree",
      $dynamicAnchor: "node",
      type: "object",
      properties: {
        data: true,
        children: {
          type: "array",
          items: { $dynamicRef: "#node" },
        },
      },
    }

    // render the component
    let root: ReactTestRenderer | undefined
    await act(async () => {
      root = create(
        <JSONSchemaViewer
          schema={fakeSchema}
          viewerOptions={{
            // To simulate custom handling of unsolved $ref
            UnresolvedRefsComponent: () => <>#node was not resolved</>,
          }}
        />,
      )
    })

    // make assertions on root
    expect(root?.toJSON()).toMatchSnapshot()
  })

  test("Overwrite default ValueComponent value", async () => {
    // tree schema, extensible
    // https://json-schema.org/draft/2020-12/release-notes#dynamicref-and-dynamicanchor
    const fakeSchema: JSONSchema = {
      $schema: "http://json-schema.org/draft-07/schema#",
      title: "CustomizationOptions",
      description: "JSON schema for customized options",
      type: "object",
      properties: {
        customField: {
          type: "string",
          description: "A customized or personalized field",
          enum: [
            "palette",
            "teddyBear",
            "tools",
            "laptop",
            "thread",
            "phone",
            "puzzle",
            "scissors",
            "hammer",
            "note",
          ],
          default: "palette",
          examples: ["tools", "note"],
        },
        customConstObject: {
          type: "object",
          const: {
            version: 5,
          },
        },
      },
      required: ["customField"],
      additionalProperties: false,
    }

    // render the component
    let root: ReactTestRenderer | undefined
    await act(async () => {
      root = create(
        <JSONSchemaViewer
          schema={fakeSchema}
          viewerOptions={{
            ValueComponent: ({ value, schema }) => {
              // render complex values as multiline JSON with 2 space indentation
              if (!["string", "number", "undefined"].includes(typeof value)) {
                return (
                  <CodeBlock language="json">{`${JSON.stringify(
                    value,
                    null,
                    2,
                  )}`}</CodeBlock>
                )
              }

              // display elementary values inline.
              const component = <code>{`${value}`}</code>

              // if schema defines a default value, ensure it is bold wherever it
              // appears (e.g. in an enum)
              if (
                typeof schema !== "boolean" &&
                schema.default &&
                value === schema.default
              ) {
                return <strong>{component}</strong>
              }

              return component
            },
          }}
        />,
      )
    })

    // make assertions on root
    expect(root?.toJSON()).toMatchSnapshot()
  })

  test("Overwrite default className value", async () => {
    const fakeSchema: JSONSchema = {
      type: "object",
      minProperties: 1,
    }

    // render the component
    let root: ReactTestRenderer | undefined
    await act(async () => {
      root = create(
        <JSONSchemaViewer schema={fakeSchema} className="jsv-custom" />,
      )
    })

    // make assertions on root
    expect(root?.toJSON()).toMatchSnapshot()
  })
})
