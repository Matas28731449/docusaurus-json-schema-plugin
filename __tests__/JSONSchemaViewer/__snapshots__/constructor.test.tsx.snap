// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSONSchemaViewer - constructor Overwrite default DescriptionComponent value 1`] = `
<details
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <div>
    <div>
      <strong>
        type
      </strong>
       : 
      <span
        style={
          {
            "opacity": "0.6",
          }
        }
      >
        object
      </span>
      <div
        style={
          {
            "marginTop": "var(--ifm-table-cell-padding)",
          }
        }
      />
      <div
        style={
          {
            "marginTop": "var(--ifm-table-cell-padding)",
          }
        }
      >
        <h1>
           
          Hello, 
          <em>
            world
          </em>
          !
           
        </h1>
      </div>
    </div>
  </div>
</details>
`;

exports[`JSONSchemaViewer - constructor Overwrite default UnresolvedRefsComponent value 1`] = `
<details
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <div>
    <div>
      <strong>
        type
      </strong>
       : 
      <span
        style={
          {
            "opacity": "0.6",
          }
        }
      >
        object
      </span>
      <ul>
        <li
          data-level={1}
          id="/properties/data"
        >
          <details
            open={false}
          >
            <summary>
              <strong>
                data
              </strong>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                any
              </span>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  any
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                >
                  Always valid
                </div>
              </div>
            </div>
          </details>
        </li>
        <li
          data-level={1}
          id="/properties/children"
        >
          <details
            open={false}
          >
            <summary>
              <strong>
                children
              </strong>
               
              (
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                any
              </span>
              ,
              ...
              )[]
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  array
                </span>
                <ul>
                  <li
                    data-level={2}
                    id="/properties/children/items"
                  >
                    <details
                      open={false}
                    >
                      <summary>
                        <code>
                          items[x]
                        </code>
                         
                        <span
                          style={
                            {
                              "opacity": "0.6",
                            }
                          }
                        >
                          any
                        </span>
                      </summary>
                      <div>
                        <div>
                          <strong>
                            type
                          </strong>
                           : 
                          <span
                            style={
                              {
                                "opacity": "0.6",
                              }
                            }
                          >
                            any
                          </span>
                          <div
                            style={
                              {
                                "marginTop": "var(--ifm-table-cell-padding)",
                              }
                            }
                          >
                            Always valid
                          </div>
                          <div
                            style={
                              {
                                "marginTop": "var(--ifm-table-cell-padding)",
                              }
                            }
                          >
                            <div>
                              #node was not resolved
                            </div>
                          </div>
                        </div>
                      </div>
                    </details>
                  </li>
                </ul>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                />
              </div>
            </div>
          </details>
        </li>
      </ul>
      <div
        style={
          {
            "marginTop": "var(--ifm-table-cell-padding)",
          }
        }
      />
    </div>
  </div>
</details>
`;

exports[`JSONSchemaViewer - constructor Overwrite default ValueComponent value 1`] = `
<details
  open={true}
>
  <summary>
    <strong>
      CustomizationOptions
    </strong>
  </summary>
  <div>
    <div>
      <strong>
        type
      </strong>
       : 
      <span
        style={
          {
            "opacity": "0.6",
          }
        }
      >
        object
      </span>
      <ul>
        <li
          data-level={1}
          id="/properties/customField"
        >
          <details
            open={false}
          >
            <summary>
              <strong>
                customField
              </strong>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                string
              </span>
               
              <strong>
                required
              </strong>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  string
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                >
                  <div>
                    <strong>
                      Possible values :
                    </strong>
                     
                    <ul>
                      <li>
                        <strong>
                          <code>
                            palette
                          </code>
                        </strong>
                      </li>
                      <li>
                        <code>
                          teddyBear
                        </code>
                      </li>
                      <li>
                        <code>
                          tools
                        </code>
                      </li>
                      <li>
                        <code>
                          laptop
                        </code>
                      </li>
                      <li>
                        <code>
                          thread
                        </code>
                      </li>
                      <li>
                        <code>
                          phone
                        </code>
                      </li>
                      <li>
                        <code>
                          puzzle
                        </code>
                      </li>
                      <li>
                        <code>
                          scissors
                        </code>
                      </li>
                      <li>
                        <code>
                          hammer
                        </code>
                      </li>
                      <li>
                        <code>
                          note
                        </code>
                      </li>
                    </ul>
                  </div>
                  <div>
                    <strong>
                      Default value :
                    </strong>
                     
                    <strong>
                      <code>
                        palette
                      </code>
                    </strong>
                  </div>
                </div>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                >
                  A customized or personalized field
                </div>
              </div>
            </div>
          </details>
        </li>
        <li
          data-level={1}
          id="/properties/customConstObject"
        >
          <details
            open={false}
          >
            <summary>
              <strong>
                customConstObject
              </strong>
               
              <span
                style={
                  {
                    "opacity": "0.6",
                  }
                }
              >
                object
              </span>
               
              <strong>
                constant
              </strong>
            </summary>
            <div>
              <div>
                <strong>
                  type
                </strong>
                 : 
                <span
                  style={
                    {
                      "opacity": "0.6",
                    }
                  }
                >
                  object
                </span>
                <div
                  style={
                    {
                      "marginTop": "var(--ifm-table-cell-padding)",
                    }
                  }
                >
                  <div>
                    <strong>
                      Constant value :
                    </strong>
                     
                    <code>
                      {
  "version": 5
}
                    </code>
                  </div>
                </div>
              </div>
            </div>
          </details>
        </li>
      </ul>
      <div
        style={
          {
            "marginTop": "var(--ifm-table-cell-padding)",
          }
        }
      >
        <div>
          🚨 
          No extra propertie(s) are authorized in this object
        </div>
      </div>
      <div
        style={
          {
            "marginTop": "var(--ifm-table-cell-padding)",
          }
        }
      >
        JSON schema for customized options
      </div>
    </div>
  </div>
</details>
`;

exports[`JSONSchemaViewer - constructor Overwrite default qualifierMessagesOrder value 1`] = `
<details
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <div>
    <div>
      <strong>
        type
      </strong>
       : 
      <span
        style={
          {
            "opacity": "0.6",
          }
        }
      >
        object
      </span>
      <div
        style={
          {
            "marginTop": "var(--ifm-table-cell-padding)",
          }
        }
      >
        <div>
          <strong>
            Length :
          </strong>
           
          <code>
            &gt;= 1 propertie(s)
          </code>
        </div>
      </div>
    </div>
  </div>
</details>
`;
